### DISCUSSION TITLE
Water points as points of attraction

### DISCUSSION THREAD

**Discussion by cherraiz** [bmcclintock/momentuHMM#165] [General]
Hi Britt and Théo. First of all, congratulations and thank you for the R package. I find it very useful and with great potential for telemetry analysis.

In my case, I want to do something more focused on simulation. I don't need very realistic trajectories, but I do want the simulated animals to visit water points in a similar frequency to how they would in reality. I am fitting the fitHMM model to real trajectories from marked individuals, using water points as activity centers and habitat type (forest, shrubland, grassland, etc.) as spatialCov. My problem is that this treats the water points as independent, so those that are far from the are not considered. But I want the simulated individuals to be attracted to the water points nearby, because the animals were marked in a specific area of the park, but I want to simulate trajectories in a greater extent, including other water points.

I have been looking at other issues and I have seen that perhaps this could be done with centroids instead of centers, in which case for each location and time the centroid is the water point which is closest to the animal. But I don't know how I could do this.

Another idea is to take the average attraction effect that the water points in the area where the original trajectory is located have, and use it as a value for all the centers in the simulation. But I don't know how to access the effect that each water point has on the model, and I think it would be less precise than the previous option.

```r
hmm_data <- prepData(data = traj,   # Real trajectories from marked individuals
                     type = 'UTM', 
                     coordNames = c('x', 'y'),
                     covNames = c('sp'),
                     spatialCovs = list(landcover = landco_rast),  # Raster with habitat
                     centers = wp_coord)  # Water points coordinates

hist(hmm_data$step, na.rm = TRUE)
hist(hmm_data$angle, na.rm = TRUE)

m <- fitHMM(data = hmm_data, 
               nbStates = 2, 
               dist = list(step = 'gamma', angle = 'wrpcauchy'),
               estAngleMean = list(angle=FALSE),
               Par0 = list(step = c(50,200,50,200), angle = c(0.7,0.1)),    # Based on the histograms
               formula = ~ 1 + landcover,
               stateNames = c('foraging', 'transit'),
)

print(mpar)

Value of the maximum log-likelihood: -3096.814 


step parameters:
----------------
     foraging  transit
mean 20.48695 134.0239
sd   15.77332 158.4810

angle parameters:
-----------------
               foraging   transit
mean          0.0000000 0.0000000
concentration 0.9788015 0.1603533

Regression coeffs for the transition probabilities:
---------------------------------------------------
                1 -> 2      2 -> 1
(Intercept) -18.539746   -3.336355
landcover1    0.000000    0.000000
landcover2   17.174538    1.067239
landcover3   18.704444   -1.572319
landcover4   49.129488 -116.849349
landcover5   87.190057  -22.271206
landcover6   15.694030    0.323918
landcover7    0.000000    0.000000
landcover8    4.889064  -98.242670

Transition probability matrix (based on mean covariate values):
---------------------------------------------------------------
           foraging   transit
foraging 0.94509680 0.0549032
transit  0.04686715 0.9531329

Initial distribution:
---------------------
    foraging      transit 
5.081506e-28 1.000000e+00 
``` 

On the other hand, when I model with one or two individuals, I have no problem, but when I include several (of the same species), I get the following error. I understand that it is a convergence problem because the variability between individuals is very high, but I am not sure about it.

```r
Error in checkInputs(nbStates, dist, Par0, estAngleMean, circularAngleMean,  : 
  Wrong number of initial parameters -- there should be 6 initial step parameters -- zero-mass parameters should be included
``` 

Thank you very much for your time and help.

Best regards,
César.

**Comment by bmcclintock** [comment]
That error is due to some of the individuals having steps of zero length. You therefore must include zero-mass parameters. See the {momentuHMM} [vignette](https://cran.r-project.org/web/packages/momentuHMM/vignettes/momentuHMM.pdf).

If you want centers of attraction or centroids to affect the movement of the individuals, then this behavior must be specified via the `DM` argument in `simData` and `fitHMM`.  See #77,  #111, and #132, and the {momentuHMM} [vignette](https://cran.r-project.org/web/packages/momentuHMM/vignettes/momentuHMM.pdf).

**Reply by cherraiz** [reply]
Hello @bmcclintock

Thank you very much for responding so quickly. I have been working on it during this time. Now I'm stuck again. I hope you can help me, but if not, thank you anyway. You have put a lot of effort into this package and all the information about it. Let me tell you a little bit... First, I tried to fit the trajectories to a recharge dynamics model like in the buffalo example in the [vignette](https://cran.r-project.org/web/packages/momentuHMM/vignettes/momentuHMM.pdf), but I found that the animal movement was not biased towards the water points (probably because the time interval between locations is too large to detect it).

Since I need the simulated animals to visit the water points, I thought it would be best to simulate a Hierarchical HMM with two levels:
Level 1: "Hydrated" or "Thirsty" based on a recharge dynamics model.
Level 2_Hydrated: "Foraging" or "Transit" based on speeds and turning angles, with parameters from a model fitted to the real trajectories.
Level 2_Thirsty: "Drinking" or "Seeking." "Seeking" with the same parameters as "Transit" but oriented towards a water point. "Drinking" when it reaches the water point and stops while recharging until Level 1 changes state to "Hydrated".
![image](https://github.com/bmcclintock/momentuHMM/assets/130759396/f71fa080-0ecb-4176-a71e-a7839c7ce322)
For this purpose, I have fitted an `"rw_mvnorm2"` model to the real trajectories, with a pseudo-matrix `DM` that includes the location (`mu.x` and `mu.y`), the velocity (`crw lag=1`), and the gradient of the distance to a water point (`D.x` and `D.y`) fixed at 0 with `fixPar`. Additionally, the `beta` model includes the land cover (`landcover`) and whether it is at a water point or not (`wp_near`, fixed to 0 in this case).

The pseudo-matrix `DM` for the simulation assigns the same values of `mu.x` and `mu.y` to the 4 states, the `crw` and `sigma` values of "Foraging" to "Foraging," and the `crw` and `sigma` values of "Transit" to "Transit" and "Seeking." For "Drinking" state, I tried to make it almost stationary, but I'm not sure if I did it correctly. "Seeking" and "Drinking" are biased towards the water points (negative value in the `D.x` and `D.y` gradients), while "Foraging" and "Transit" are not (perhaps I will included a slight repulsion effect like in #108).

I want the `beta` in "Hydrated" to be the one from the fitted "Foraging" and "Transit" model, and the `beta` in "Thirsty" to be unable to change from "Drinking" to "Seeking" and to change compulsorily from "Seeking" to "Drinking" upon entering a water point `wp_near`. 

I feel like I'm heading in the right direction, but I'm getting errors related to parameter input. `Beta` tells me it should be a 3x2 matrix (there are 3 variables, but one is categorical with 3 categories). I'm also not providing `Delta` correctly, I think because I'm not providing it on the working scale. Maybe you can help me with this.

Additionally, I would like to know if I could obtain average speeds and some measure of angles for each state from the `"rw_mvnorm2"` model that I fitted to the real trajectories.

Regarding the number of observations per level in `SimHierData`, I think I should use the same value for both levels, but I'm not sure.

Here is the code and the [data](https://drive.google.com/file/d/1xLLXda-hwBJXUsUUruIXJv-K9c3PCoJR/view?usp=sharing) (in Google Drive), which includes the crawled trajectories and the spatial covariates.

Thank you very much for the help. 
Best regards. 

```r
# 1. Data object for momentuHMM modeling
spatialCovs <- list(wp_intercept = r_stack$intercept,
                    wp_near = r_stack$near_wp,
                    wp_dist = wp_dist,
                    D.x = dist_scaled$rast.grad.x,
                    D.y = dist_scaled$rast.grad.y,
                    landcover = landco_rast)
hmmData <- prepData(crw_traj, spatialCovs = spatialCovs, altCoordNames = "mu")
head(hmmData)

# 2. Fitting the model to real trajectories
DM <- list(mu=matrix(c("mu.x_tm1","crw(mu.x_tm1,lag=1)","D.x",                    0,0,0,0,
                       "mu.x_tm1",                    0,"D.x","crw(mu.x_tm1,lag=1)",0,0,0,
                       "mu.y_tm1","crw(mu.y_tm1,lag=1)","D.y",                    0,0,0,0,
                       "mu.y_tm1",                    0,"D.y","crw(mu.y_tm1,lag=1)",0,0,0,
                                0,                    0,    0,                    0,1,0,0,
                                0,                    0,    0,                    0,0,1,0,
                                0,                    0,    0,                    0,0,0,1,
                                0,                    0,    0,                    0,0,0,1,
                                0,                    0,    0,                    0,1,0,0,
                                0,                    0,    0,                    0,0,1,0
                       ),2*5,7,byrow=TRUE,
                        dimnames = list(c(paste0("mean.x_",1:2),paste0("mean.y_",1:2),
                                          paste0("sigma.x_",1:2),paste0("sigma.xy_",1:2),
                                          paste0("sigma.y_",1:2)),
                                        c("mu_tm1","crw_1","D","crw_2","sigma_1","sigma_2",
                                          "sigma.xy"))))

Par0 <- list(mu=c(1,0.5,0,0.1,log(50000),log(50000000),0))
names(Par0$mu) <- c("mu.x_tm1_1","crw(mu.x_tm1,lag=1)_1","D.x_1","crw(mu.x_tm1,lag=1)_2",
                    "sigma.x_1","sigma.x_2","sigma.xy_1")
fixPar=list(mu=c(1,NA,0,NA,NA,NA,0),
            delta = c(0.1,0.9),
            beta=matrix(c(NA,NA,
                          NA,NA,
                          NA,NA,
                          NA,NA,
                          0,0),5,byrow = TRUE))

hmmFit <- fitHMM(hmmData, 
                 nbStates = 2, 
                 dist = list(mu = "rw_mvnorm2"),
                 DM = DM,
                 estAngleMean = list(angle=FALSE),
                 Par0 = Par0,
                 fixPar = fixPar,
                 formula = ~ 1 + landcover +wp_near,
                 stateNames = c('foraging', 'transit'), fit = TRUE,
                 mvnCoords="mu"); hmmFit
# plot(hmmFit)

# 3. Hier simulation
# 3.1. Hier states 
hierStates <- data.tree::Node$new("States")
hierStates$AddChild("Hydrated")
hierStates$Hydrated$AddChild("Foraging", state=1)
hierStates$Hydrated$AddChild("Transit", state=2)
hierStates$AddChild(name="Thirsty")
hierStates$Thirsty$AddChild("Drinking", state=3)
hierStates$Thirsty$AddChild("Seeking", state=4)
# plot(hierStates)

# 3.2. Hier distribution
hierDist <- data.tree::Node$new("dist")
hierDist$AddChild("level1") 
hierDist$AddChild("level2")
hierDist$level2$AddChild("mu", dist = "rw_mvnorm2")

# 3.3. Hier formula
hierFormula <- data.tree::Node$new("Formula")
hierFormula$AddChild("level1", formula=~recharge(g0=~1,theta=~wp_near))
hierFormula$AddChild("level2", formula=~1 + landcover + wp_near)

# 3.4. Hier beta
hierBeta <- data.tree::Node$new("Beta")
hierBeta$AddChild("level1",
                  beta=matrix(c(0,-1,0,-1),2,2),g0=-0.3,theta=c(-0.25,5))
hierBeta$AddChild("level2")
hierBeta$level2$AddChild("Hydrated",
                         beta=getPar(hmmFit)$beta)
hierBeta$level2$AddChild("Thirsty",
                         beta=matrix(c(-100,-100,
                                       -100,-100,
                                       -100,-100,
                                       -100,-100,
                                       -100, 100), 
                                     nrow=5,
                                     ncol=2, byrow=TRUE))

# 3.5. Hier delta
hierDelta <- data.tree::Node$new("Delta")
hierDelta$AddChild("level1",delta=matrix(c(0.5, 0.5),1))
hierDelta$AddChild("level2")
hierDelta$level2$AddChild("Hydrated",delta=matrix(getPar0(hmmFit)$delta, 1))
hierDelta$level2$AddChild("Thirsty",delta=matrix(c(0, 1),1))

# 3.6. DM
DM <- list(mu=matrix(c("mu.x_tm1","crw(mu.x_tm1,lag=1)","D.x",    0,                    0,                    0,0,0,0,0,
                       "mu.x_tm1",                    0,"D.x",    0,"crw(mu.x_tm1,lag=1)",                    0,0,0,0,0,
                       "mu.y_tm1",                    0,    0,"D.x",                    0,"crw(mu.x_tm1,lag=1)",0,0,0,0,
                       "mu.y_tm1",                    0,    0,"D.x","crw(mu.y_tm1,lag=1)",                    0,0,0,0,0,
                       "mu.x_tm1","crw(mu.x_tm1,lag=1)","D.y",    0,                    0,                    0,0,0,0,0,
                       "mu.x_tm1",                    0,"D.y",    0,"crw(mu.y_tm1,lag=1)",                    0,0,0,0,0,
                       "mu.y_tm1",                    0,    0,"D.y",                    0,"crw(mu.y_tm1,lag=1)",0,0,0,0,
                       "mu.y_tm1",                    0,    0,"D.y","crw(mu.y_tm1,lag=1)",                    0,0,0,0,0,
                                0,                    0,    0,    0,                    0,                    0,1,0,0,0,
                                0,                    0,    0,    0,                    0,                    0,0,1,0,0,
                                0,                    0,    0,    0,                    0,                    0,0,0,1,0,
                                0,                    0,    0,    0,                    0,                    0,0,1,0,0,
                                0,                    0,    0,    0,                    0,                    0,0,0,0,1,
                                0,                    0,    0,    0,                    0,                    0,0,0,0,1,
                                0,                    0,    0,    0,                    0,                    0,0,0,0,1,
                                0,                    0,    0,    0,                    0,                    0,0,0,0,1,
                                0,                    0,    0,    0,                    0,                    0,1,0,0,0,
                                0,                    0,    0,    0,                    0,                    0,0,1,0,0,
                                0,                    0,    0,    0,                    0,                    0,0,0,1,0,
                                0,                    0,    0,    0,                    0,                    0,0,1,0,0
                       ),4*5,10,byrow=TRUE,
                     dimnames = list(c(paste0("mean.x_",1:4),paste0("mean.y_",1:4),paste0("sigma.x_",1:4),
                                       paste0("sigma.xy_",1:4),paste0("sigma.y_",1:4)),
                                     c("mu_tm1","crw_1","D_12", "D_34","crw_24","crw_3","sigma_1","sigma_24",
                                       "sigma_3","sigma.xy"))))

# 3.7. Par
Par0 <- getPar0(hmmFit)
Par <- list(mu=c(Par0$Par$mu[1:2], D_12 = 0, D_34=-100, crw_24=Par0$Par$mu[[4]], crw_3=0.9, Par0$Par$mu[5],
                 sigma_24=Par0$Par$mu[[6]],sigma_3=log(1),Par0$Par$mu[7]))

# 3.8. Initial positions
posits.df <- split(matrix(c(723904,723589,726974,723929,725284,4097746,4096536,4095861,4097626,4100256), 
                          nrow = 5, ncol = 2), 1:5)

# 3.9. Simulation
obsPerLevel <- data.tree::Node$new("simHierData")
obsPerLevel$AddChild("level1",obs=500)
obsPerLevel$AddChild("level2",obs=500)
hmmSim <- simHierData(nbAnimals = 5, hierStates = hierStates, hierDist = hierDist, 
                      Par = Par, hierBeta = hierBeta, hierDelta = hierDelta, 
                      hierFormula = hierFormula, spatialCovs = spatialCovs, DM = DM, 
                      states = TRUE, initialPosition = posits.df, mvnCoords = "mu",
                      obsPerLevel = obsPerLevel)
```
